user/syscall_test.c

```diff
@@ -0,0 +1,25 @@
+#include <stddef.h>
+
+int write(int fd, const char *buf, size_t count) {
+  int retval, errval;
+  asm volatile("move $a0, %2\n"
+               "move $a1, %3\n"
+               "move $a2, %4\n"
+               "li $v0, 5\n" /* SYS_WRITE */
+               "syscall\n"
+               "move %0, $v0\n"
+               "move %1, $v1\n"
+               : "=r"(retval), "=r"(errval)
+               : "r"(fd), "r"(buf), "r"(count)
+               : "4", "5", "6", "2", "3");
```

Comment:
I admit, I am not exactly confident how exactly these asm inlines work. What I intended to express is that the compiler shouldn't use registers `a0`, `a1`, `a2`, `v0`, and `v1` for input/output operands, because I use these registers internally (otherwise the compiler placed `count` in `$a0`, so the first assembly instruction would overwrite it...). It didn't accept `"$a0"` nor `"a0"` as input, so I used register numbers, and then it immediately started working as I intended - `fd` got placed in `$a3`, `buf` in `$t0` and `count` in `$t1`.

I suppose there may be a more elegant way to do this, but the GCC inline asm documentation confuses me a lot. I would even prefer to write this in pure asm.